tower-util/src/combinators/with.rs

```diff
@@ -0,0 +1,70 @@
+use std::marker::PhantomData;
+use std::task::{Context, Poll};
+use tower_layer::Layer;
+use tower_service::Service;
+
+pub struct With<S, F> {
```

Comment:
`tower-http` contains a [`tower-request-modifier` crate](https://github.com/tower-rs/tower-http/tree/master/tower-request-modifier) that does a bunch of HTTP-specific request mutations...I wonder if that could be reimplemented on top of this API, eventually? It would be nice to keep that in mind and make sure this is possible.

Reply:
I'll give https://github.com/tower-rs/tower-http/tree/master/tower-request-modifier a fork and try it out when I get some spare time.

Reply:
https://github.com/tower-rs/tower-http/pull/31

I departed from the previous API by a fair bit but I think it turned out ok.